### PR TITLE
Parallelize send_chain_information_for_senders

### PR DESCRIPTION
**Pull request by andresilva91**
## Motivation

Inspired by #2822, saw that we could do the same parallelization for `send_chain_information_for_senders` as well

## Proposal

Parallelize sending `send_chain_information`

## Test Plan

CI

## Release Plan

- Nothing to do / These changes follow the usual release cycle.

### CODE REVIEW

**Comment by andresilva91** [comment]
* **#2848** <a href="https://app.graphite.dev/github/pr/linera-io/linera-protocol/2848?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @andresilva91 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: linera-core/src/updater.rs [hunk 1]
```diff
@@ -379,17 +379,29 @@ where
...
+                    updater
+                        .send_chain_information(
+                            sender,
+                            next_height,
+                            CrossChainMessageDelivery::Blocking,
+                        )
+                        .await
+                }
+            })
+            .buffer_unordered(self.chain_worker_count);
```
[the remark this answers is not in the document]
Comment (afck) [reply linera-core/src/updater.rs]: In general I agree, but I think in this case we already computed `self.chain_worker_count` when we created the updater, and divided by the number of validators.

I don't really know what's the best alternative here: Last time this was discussed we decided against changing how chain worker actors are evicted. This maybe needs further discussion, but is beyond the scope of this PR?

File: linera-core/src/updater.rs [hunk 2]
```diff
@@ -296,28 +296,18 @@ where
                     let mut chain_heights = BTreeMap::new();
                     for certificate in certificates {
                         let block_chain_id = certificate.value().chain_id();
-                        let block_height = certificate.value().height();
+                        let block_height = certificate.value().height().try_add_one()?;
```
Comment (deuszx) [inline linera-core/src/updater.rs:299]: This is changing the logic - is it still correct?

Comment (afck) [reply linera-core/src/updater.rs:299]: There was a `try_add_one` in line 309 before. That's now gone, so we have to add one here, if I understand correctly.

Comment (andresilva91) [reply linera-core/src/updater.rs:299]: Yes, what @afck said. I just did the height increment earlier so we can call `send_chain_information_to_chain_heights` and reuse the code

Comment (deuszx) [reply linera-core/src/updater.rs:299]: We might modify the entry in `chain_heights` prematurely though – previously the `block_height.max(*h)` might not modify `chain_heights` b/c the previously-stored `h` was higher. Now we'd compare it on the `block_height + 1`.

**Comment by andresilva91** [comment]
### Merge activity

* **Nov 8, 9:32 AM EST**: A user started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/linera-io/linera-protocol/2848).

File: linera-core/src/updater.rs [hunk 3]
```diff
@@ -370,6 +360,25 @@ where
         Ok(())
     }
 
+    async fn send_chain_information_to_chain_heights(
```
Comment (ma2bd) [inline linera-core/src/updater.rs:363]: @andresilva91 @afck What is this name supposed to mean?

Comment (ma2bd) [reply linera-core/src/updater.rs:363]: `send_chain_information_up_to_chain_heights`?

Comment (afck) [reply linera-core/src/updater.rs:363]: Good point, it should be "up to".

And maybe just `send_chain_info_up_to_heights`? No need to repeat "chain". And we already use the abbreviation "info" in the type itself.

Comment (andresilva91) [reply linera-core/src/updater.rs:363]: Good point! I can send a PR renaming it on Monday

Comment (ndr-ds) [reply linera-core/src/updater.rs:363]: https://github.com/linera-io/linera-protocol/pull/2859

**Pull request merged**